### PR TITLE
improve monitor performance

### LINKED ISSUE #10347: monitor: consul monitor and consul debug are missing many logs
#### Overview of the Issue

`consul monitor` (and `consul debug` which uses the same monitoring) are supposed to show all of the consul logs. Instead they appear to only have logs for some loggers and not others. This problems exists on master, and I believe has been a problem for the last few major releases.

I suspect the issue may actually be in  https://github.com/hashicorp/go-hclog/, but filing the issue here in case it is only related to Consuls implementation.

Edit: removed the reproduction steps, which were not correct

### PR DESCRIPTION
**Pull request by dhiaayachi**
Fixes #10347

improve monitor performance to avoid filling the log channel and loose logs when running monitor or debug.

An issue was reported that we loose some logs when are running `consul monitor` or `consul debug` commands. 

the current underlying monitor implementation avoid blocking when writing logs to not impact performance and allow losing some logs when we overflow the log channel (size 512 lines of code). This PR is to enhance the performance of the log channel read so we can avoid loosing logs.

### CODE REVIEW

File: agent/agent_endpoint.go
Comment (dnephin) [inline agent/agent_endpoint.go:1222]: `AfterFunc` calls `flusher.Flush` in a separate goroutine, which I'm not sure is necessarily safe. It also creates an unnecessary channel each time.

What do you think about using just a `time.Time` here?

```go
lastFlush := time.Now()

...

if time.Since(lastFlush) > flushDelay {
    flusher.Flush()
    lastFlush = time.Now()
}


```

**Review by dnephin — COMMENTED** [review]
Thanks! I think this is going to be a good improvement. Left some suggestions below.

I also had a question about 1eeddf7ad801cb8f55cd088c275e1a59bd126209. Generally it is expected that once something is closed or Stopped it can't be started again. Was that behaviour causing a problem? Maybe we could document that a monitor is only safe for one use, instead of changing that behaviour?

**Review by dhiaayachi — COMMENTED** [review]
> Thanks! I think this is going to be a good improvement. Left some suggestions below.
> 
> I also had a question about [1eeddf7](https://github.com/hashicorp/consul/commit/1eeddf7ad801cb8f55cd088c275e1a59bd126209). Generally it is expected that once something is closed or Stopped it can't be started again. Was that behaviour causing a problem? Maybe we could document that a monitor is only safe for one use, instead of changing that behaviour?

Yes the issue was that if we stop the monitor we can't start it again. I was testing and I came cross that issue. 
Why do we not expect the monitor to be started multiple times. This is in the server side so it's a plausible scenario to start and stop monitoring multiple times to debug an agent. am I missing something?

File: agent/agent_endpoint.go
Comment (dnephin) [reply agent/agent_endpoint.go:1222]: That is true, it does mean that when logs are infrequent we wouldn't get the last few logs until the monitor is closed. Since this is a debugging interface maybe that is ok? Generally a monitor is only going to be run for a few seconds or minutes. So even if we don't write logs often, when the user closes the monitor they should get the last few logs, right?

File: .changelog/10368.txt
Comment (dnephin) [reply .changelog/10368.txt:2]: A tool that finds the right PR number by using the github API and creates the right filename and prompts for the type would be pretty cool!

File: logging/monitor/monitor.go
Comment (dnephin) [reply logging/monitor/monitor.go:106]: Ya, this is a good find! This should help a lot I think

File: agent/agent_endpoint.go
Comment (dhiaayachi) [reply agent/agent_endpoint.go:1222]: The challenge is more with the tests, we will need to rework a lot of tests for this. As we use the log interceptor to check if we got the logs we expect and once we stop the monitor the interceptor is not anymore available to check the remaining of the logs.

File: .changelog/10368.txt
Comment (dhiaayachi) [reply .changelog/10368.txt:2]: yes that's exactly what I did :) I will be publishing it soon. I just wanted to see if go-changelog would accept it as part of the project or I need to publish it separately

File: agent/agent_endpoint.go
Comment (dhiaayachi) [reply agent/agent_endpoint.go:1222]: Added complexity is that I noticed the api tests fail on the pipeline but not locally when we miss/remove the flush. I think because locally we use localhost as interface so the flush is implicit.

Comment (dhiaayachi) [reply agent/agent_endpoint.go]: I changed it to use a ticker, it do not avoid using extra resources but it solve the race issue. IMO we should avoid having data not flushed until we close the connection. The user experience is not ideal.

File: logging/monitor/monitor.go
Comment (dnephin) [inline logging/monitor/monitor.go:38]: Your question over in https://github.com/hashicorp/consul/pull/9285#discussion_r650157725 made me think of this. I think this is actually not safe for the reason I outline there.

This needs to be a broadcast, because multiple goroutines need to be informed of the change (both the `Start` goroutine, and   whatever is calling `Write`). 

If we need to change a bunch of tests to account for the flushing change let me know, maybe I can help with that.

File: agent/agent_endpoint.go
Comment (dnephin) [inline agent/agent_endpoint.go:1224]: Nice! I think I've seen this pattern work well before. This should work well to periodically flush.

File: agent/agent_endpoint.go
Comment (dnephin) [inline agent/agent_endpoint.go:1207]: The docs for NewTicker say:

> Stop the ticker to release associated resources.

so I think we need to do this to prevent leaking some things on each monitor:

```suggestion
	const flushDelay = 200 * time.Millisecond
	flushTicker := time.NewTicker(flushDelay)
	defer flushTicker.Stop()
```

File: logging/monitor/monitor.go
Comment (dhiaayachi) [reply logging/monitor/monitor.go]: I reverted the change, anyway after reading the code closely it's not meant to be concurrent (we create a new monitor in the handler for each request, reusing the same monitor will not work on both start and stop)

File: logging/monitor/monitor.go
Comment (dnephin) [reply logging/monitor/monitor.go:102]: I guess we still register the sink without knowing if the goroutine has started. We could create a `chStarted` and close it from the goroutine, and block on it from the main goroutine, to further delay the `RegisterSink` and be confident that the goroutine has started. 

Would also be fine as a follow up.

**Review by dnephin — APPROVED** [review]
LGTM! One idea for a small improvement, but not blocking

**Comment by dhiaayachi** [comment]
@dnephin I added the suggested fix but went with a `WaitGroup` as it seems appropriate for this use case, any thought?

**Comment by dnephin** [comment]
That works. I think it could potentially be confusing the the reader since generally wait groups are used for goroutines to end (not start), but functionally I think it's going to do the right thing.

**Comment by hc-github-team-consul-core** [comment]
:cherries: If backport labels were added before merging, cherry-picking will start automatically.

To retroactively trigger a backport after merging, add backport labels and re-run https://circleci.com/gh/hashicorp/consul/386958.

**Comment by hc-github-team-consul-core** [comment]
:cherries: If backport labels were added before merging, cherry-picking will start automatically.

To retroactively trigger a backport after merging, add backport labels and re-run https://circleci.com/gh/hashicorp/consul/386967.

**Comment by hc-github-team-consul-core** [comment]
:cherries::white_check_mark: Cherry pick of commit c8ba2d40fd8d6915c3e469bfa5bb252f7ab31b70 onto `release/1.10.x` succeeded!

**Comment by hc-github-team-consul-core** [comment]
:cherries::white_check_mark: Cherry pick of commit c8ba2d40fd8d6915c3e469bfa5bb252f7ab31b70 onto `release/1.9.x` succeeded!

**Comment by hc-github-team-consul-core** [comment]
:cherries::white_check_mark: Cherry pick of commit c8ba2d40fd8d6915c3e469bfa5bb252f7ab31b70 onto `release/1.8.x` succeeded!

**Pull request merged**
